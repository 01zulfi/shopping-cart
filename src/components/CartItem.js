import React, { useState } from "react";
import cart from "../CartManager";

const CartItem = ({
  name,
  price,
  id,
  quantity,
  setTotalHandler,
  setCartLength,
}) => {
  const [stateQuantity, setStateQuantity] = useState(quantity);
  const [spanMessage, setSpanMessage] = useState("");

  const decrementQuantity = () => {
    if (stateQuantity === 1) return;
    setSpanMessage("");
    cart.decrementQuantity(id);
    setTotalHandler(cart.total);
    setCartLength(cart.cartLength);
    setStateQuantity(cart.find(id).quantity);
  };

  const incrementQuantity = () => {
    setSpanMessage("");
    cart.incrementQuantity(id);
    setTotalHandler(cart.total);
    setCartLength(cart.cartLength);
    setStateQuantity(cart.find(id).quantity);
  };

  const setQuantity = (event) => {
    if (event.target.value < 1) {
      setSpanMessage("Quantity must be greater than 1");
      return;
    }

    setSpanMessage("");
    cart.setQuantity(id, event.target.value);
    setTotalHandler(cart.total);
    setCartLength(cart.cartLength);
    setStateQuantity(cart.find(id).quantity);
  };

  return (
    <div className="cart-item-div" key={id}>
      <p>{name}</p>
      <p>{price}</p>
      <button className="decrement-button" onClick={decrementQuantity}>
        -
      </button>
      <label htmlFor="quantity-input" className="quantity-input">
        Quantity:
        <input
          type="number"
          id="quantity-input"
          onChange={setQuantity}
          value={stateQuantity}
        />
      </label>
      <span>{spanMessage}</span>
      <button className="increment-button" onClick={incrementQuantity}>
        +
      </button>
    </div>
  );
};

export default CartItem;
