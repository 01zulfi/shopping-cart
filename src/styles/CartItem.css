div.cart-item-div {
  display: flex;
  flex-direction: column;
}

div.name-and-price {
  display: flex;
  justify-content: space-between;
  font-size: 19px;
}

div.quantity-div {
  display: flex;
  align-self: center;
  justify-content: space-between;
  align-items: center;
  gap: 5px;
}

div.quantity-div button {
  padding: 5px;
}

.quantity-input {
  text-align: center;
}

.quantity-input input {
  background-color: #d4bff9;
  border: 0;
}
