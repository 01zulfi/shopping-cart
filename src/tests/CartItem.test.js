import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import CartItem from "../components/CartItem";

test("cart item renders", () => {
  const { container } = render(<CartItem />);
  expect(container).toMatchSnapshot();
  expect(container.firstChild).toHaveClass("cart-item-div");
});

test("diplays correct quantity value", () => {
  const { container } = render(<CartItem price={1} quantity={2} />);
  const input = container.querySelector("input");
  expect(Number(input.value)).toBe(2);
});
