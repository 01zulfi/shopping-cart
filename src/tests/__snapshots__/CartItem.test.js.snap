// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cart item renders 1`] = `
<div>
  <div
    class="cart-item-div"
  >
    <p />
    <p />
    <button
      class="decrement-button"
    >
      -
    </button>
    <label
      class="quantity-input"
      for="quantity-input"
    >
      Quantity:
      <input
        id="quantity-input"
        type="number"
        value=""
      />
    </label>
    <span />
    <button
      class="increment-button"
    >
      +
    </button>
  </div>
</div>
`;
